docs/_build/html/_sources/tiltfile_concepts.md.txt

```diff
@@ -0,0 +1,106 @@
+```python
+text = local('./foo.py') # runs command foo.py
+k8s_yaml(text)
+```
+
+`local` runs a command, and returns its `stdout` as a Blob. (A Blob is a string but is interpreted by `k8s_yaml` as text instead of as a filename.) Note: Tilt doesn't know what files a command accesses, so you need to use the function `read_file` to record accesses. If you don't call `read_file`, Tilt won't reexecute the `Tiltfile` when those files change. For example, if `foo.py` depends on the files `config/base.yaml` and `data/versions.txt`:
+
+```python
+read_file('config/base.yaml')
+read_file('data/versions.txt')
+text = local('./foo.py')
+k8s_yaml(text)
+```
+
+You can also use python features like list comprehensions. For example, if you have a script that generates yaml for one microservice at a time, you could do:
+
+```python
+# define a function that returns the config for one microservice
+def microservice_config(name):
+  # record file access, using python string substitution to generate filename
+  read_file('config/%s.yaml' % name)
+  # run the script with an argument
+  return local('./config/generate.py %s' % name)
+
+# define the service names
+services = ['frontend', 'backend', 'users', 'graphql']
+
+# loop over each service and register its config
+[k8s_yaml(microservice_config(service)) for service in services]
+```
+
+Using `local` judiciously can let you use existing tools with Tilt, without having to rewrite or abandon them immediately.
+
+## Build
+The `docker_build` function aims to support most usages of docker. Here's a cheat-sheet that maps docker command lines to a `docker_build` call:
+
+```python
+# docker build -t companyname/frontend ./frontend
+docker_build("companyname/frontend", "frontend")
+
+# docker build -t companyname/frontend -f frontend/Dockerfile.dev frontend
+docker_build("companyname/frontend", "frontend", dockerfile="frontend/Dockerfile.dev")
+
+# docker build -t companyname/frontend --build-arg target=local frontend
+docker_build("companyname/frontend", "frontend", build_args={"target": "local"})
+```
+
+You can combine multiple optional arguments.
+
+## Resources
+Tilt's UI makes it easier to find errors by grouping related status and output. E.g., when you edit a file, you want to know what error it caused, whether it's an error at build-time, deploy-time, or run-time. Tilt calls these groupings "Resources". Each Resource has a line in the UI that can be expanded and investigated.
+
+Tilt generates these groups after executing your `Tiltfile`. We're actively working on how to group in ways that make the most intuitive sense, so the specific algorithm is in-flux. We'll expand this paragraph when it's more settled.
+
+You can configure a resource with a call to `k8s_resource`. Today, the only relevant configuration is the argument `port_forwards`. Tilt supports a few alternatives:
```

Comment:
worth it to mention how we currently name resources/that you can rename them with this func?

Reply:
I think that's very likely to change, and if people need that we can help them on an ad hoc basis.